src/hotspot/share/opto/node.cpp

```diff
@@ -1696,6 +1696,670 @@ bool Node::add_to_worklist(Node* n, Node_List* worklist, Arena* old_arena, Vecto
+  }
+}
+
+void PrintBFS::print_node_idx(Node* n) {
+  Compile* C = Compile::current();
+  char buf[30];
+  if (n == nullptr) {
+    sprintf(buf,"_");           // null
+  } else if (C->node_arena()->contains(n)) {
+    sprintf(buf, "%d", n->_idx);  // new node
+  } else {
+    sprintf(buf, "o%d", n->_idx); // old node
+  }
+  tty->print("%6s", buf);
+}
+
+void PrintBFS::print_node_block(Node* n) {
+  Compile* C = Compile::current();
+  Block* b = C->node_arena()->contains(n)
+             ? C->cfg()->get_block_for_node(n)
+             : nullptr; // guard against old nodes
+  if (b == nullptr) {
+    tty->print("     _");
+    tty->print("     _");
+    tty->print("   _");
+  } else {
+    print_node_idx(b->head());
+    if (b->_idom) {
+      print_node_idx(b->_idom->head());
+    } else {
+      tty->print("     _");
+    }
+    tty->print("%4d", b->_dom_depth);
+  }
+}
+
+// filter, and add to worklist, add info, note traversal edges
+void PrintBFS::maybe_traverse(Node* src, Node* dst) {
+  if (dst != nullptr &&
+     (filter_category(dst, _filter_visit) ||
+      filter_category(dst, _filter_boundary) ||
+      dst == _start)) { // correct category or start?
+    if (find_info(dst) == nullptr) {
+      // never visited - set up info
+      _worklist.push(dst);
+      int d = 0;
+      if (dst != _start) {
+        d = find_info(src)->distance() + 1;
+      }
+      make_info(dst, d);
+    }
+    if (src != dst) {
+      // traversal edges useful during select
+      find_info(dst)->edge_bwd.push(src);
+    }
+  }
+}
+
+void PrintBFS::print_header() const {
+  tty->print("   d");                         // distance
+  if (_all_paths) {
+    tty->print(" apd");                       // all paths distance
+  }
+  if (_print_blocks) {
+    tty->print(" [head  idom  d]");           // block
+  }
+  if (_print_old) {
+    tty->print("   old");                     // old node
+  }
+  tty->print(" dump\n");                      // node dump
+  tty->print("---------------------------------------------\n");
+}
+
+void PrintBFS::print_node(Node* n) {
+  tty->print("%4d", find_info(n)->distance());// distance
+  if (_all_paths) {
+    Info* info = find_info(n);
+    int apd = info->distance() + info->distance_from_target();
+    tty->print("%4d", apd);                   // all paths distance
+  }
+  if (_print_blocks) {
+    print_node_block(n);                      // block
+  }
+  if (_print_old) {
+    print_node_idx(old_node(n));              // old node
+  }
+  tty->print(" ");
+  n->dump("\n", false, tty, &_dcc);           // node dump
+}
+
+//------------------------------dump_bfs--------------------------------------
+// Call this from debugger
+// Useful for BFS traversal, shortest path, all path, loop detection, etc
+// Designed to be more readable, and provide additional info
+// To find all options, run:
+//   find_node(0)->dump_bfs(0,0,"H")
+void Node::dump_bfs(const int max_distance, Node* target, char const* options) {
```

Comment:
Would be great to have a short version `Node::dump_bfs(n)` equivalent to `Node::dump_bfs(n, 0, 0)`, for convenience.